server/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java

```diff
@@ -829,20 +842,27 @@ private static void parseNonDynamicArray(
 
         XContentParser parser = context.parser();
         XContentParser.Token token;
+        int elements = 0;
         while ((token = parser.nextToken()) != XContentParser.Token.END_ARRAY) {
             if (token == XContentParser.Token.START_OBJECT) {
+                elements = Integer.MAX_VALUE;
                 parseObject(context, lastFieldName);
             } else if (token == XContentParser.Token.START_ARRAY) {
+                elements = Integer.MAX_VALUE;
                 parseArray(context, lastFieldName);
             } else if (token == XContentParser.Token.VALUE_NULL) {
                 parseNullValue(context, lastFieldName);
```

Comment:
What about null value? Should we do `elements++` in this case?

Reply:
Agreed, we should count nulls. It is possible that this field has `null_value` configured and therefore if we skip nulls, we lose such values on reindex. In regular synthetic source we produce such values because nulls get replaced with appropriate value during indexing.

Reply:
Good call, added along with a test.